### ISSUE TITLE
Neural Networks

### ISSUE THREAD

**Issue by Leeevai** [Leeevai/Rustorch#3]
# Implement Neural Network Support

## Summary

Add support for basic neural networks using the `matrix` crate by introducing two key structs:

- `Layer<T, A>`: Represents a single fully connected layer with weights, biases, and an activation function.
- `NeuralNetwork<T, A>`: Represents a feedforward neural network composed of multiple layers.

This will enable users to define, run, and benchmark neural networks efficiently using both sequential and concurrent operations.

---

## Motivation

The `matrix` crate already provides powerful and efficient matrix operations with optional concurrency. Leveraging this infrastructure to support neural networks would:

- Extend the library's capabilities into machine learning.
- Provide a performant and ergonomic way to experiment with forward propagation.
- Demonstrate real-world use cases of concurrent matrix computation.

---

### `Layer`

```rust
pub struct Layer<T, A>
where
    T: Default + Copy + Clone + Send + Sync,
    A: ActivationFunction<T>,
{
    pub weights: Matrix<T>,
    pub biases: Matrix<T>,
    pub activation: A,
    pub concurrent: bool,
}
```

- `weights`: Matrix of shape `(output_size, input_size)`.
- `biases`: Matrix of shape `(output_size, 1)`.
- `activation`: Generic activation function implementing `ActivationFunction<T>`.
- `concurrent`: Whether to use concurrent matrix operations.

### `NeuralNetwork`

```rust
pub struct NeuralNetwork<T, A>
where
    T: Default + Copy + Clone + Send + Sync,
    A: ActivationFunction<T>,
{
    pub layers: Vec<Layer<T, A>>,
}
```

- `layers`: A vector of `Layer`s stacked to form the neural network.

---

### `ActivationFunction<T>`

```rust
pub trait ActivationFunction<T> {
    fn activate(&self, x: T) -> T;
}
```

Additional methods like `derivative()` could be added later for training support (e.g. backpropagation).

---

## Minimal Example

```rust
let layer = Layer {
    weights: Matrix::<f64>::random(3, 2)?,
    biases: Matrix::<f64>::zeros(3, 1)?,
    activation: ReLU,
    concurrent: true,
};

let network = NeuralNetwork {
    layers: vec![layer],
};
```

---

## Other Features

- Backpropagation and gradient descent for training.
- Support for different activation functions (ReLU, Sigmoid, Tanh, etc.).
- Serialization of network weights and structure.
- Integration with performance benchmarking tools already in place.

---

## Related

- Builds on top of `Matrix<T>` APIs (e.g., dot product, transpose, matrix multiplication).
- Expands test suite with machine learning examples and timing comparisons.

---

## Priority

**High** – This significantly increases the practical use and visibility of the `matrix` crate.
